hazelcast-client-new/src/main/java/com/hazelcast/client/proxy/ClientRingbufferProxy.java

```diff
@@ -0,0 +1,248 @@
+import com.hazelcast.client.impl.protocol.codec.RingbufferReadOneCodec;
+import com.hazelcast.client.impl.protocol.codec.RingbufferRemainingCapacityCodec;
+import com.hazelcast.client.impl.protocol.codec.RingbufferSizeCodec;
+import com.hazelcast.client.impl.protocol.codec.RingbufferTailSequenceCodec;
+import com.hazelcast.client.spi.ClientProxy;
+import com.hazelcast.client.spi.impl.ClientInvocation;
+import com.hazelcast.client.spi.impl.ClientInvocationFuture;
+import com.hazelcast.client.util.ClientDelegatingFuture;
+import com.hazelcast.core.ICompletableFuture;
+import com.hazelcast.core.IFunction;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.nio.serialization.SerializationService;
+import com.hazelcast.ringbuffer.OverflowPolicy;
+import com.hazelcast.ringbuffer.ReadResultSet;
+import com.hazelcast.ringbuffer.Ringbuffer;
+import com.hazelcast.ringbuffer.impl.client.PortableReadResultSet;
+import com.hazelcast.util.ExceptionUtil;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import static com.hazelcast.ringbuffer.impl.RingbufferProxy.MAX_BATCH_SIZE;
+import static com.hazelcast.util.Preconditions.checkFalse;
+import static com.hazelcast.util.Preconditions.checkNotNegative;
+import static com.hazelcast.util.Preconditions.checkNotNull;
+import static com.hazelcast.util.Preconditions.checkTrue;
+
+public class ClientRingbufferProxy<E> extends ClientProxy implements Ringbuffer<E> {
+
+    private static final ClientMessageDecoder ADD_ASYNC_ASYNC_RESPONSE_DECODER = new ClientMessageDecoder() {
+        @Override
+        public Long decodeClientMessage(ClientMessage clientMessage) {
+            return RingbufferAddCodec.decodeResponse(clientMessage).response;
+        }
+    };
+
+    private static final ClientMessageDecoder ADD_ALL_ASYNC_RESPONSE_DECODER = new ClientMessageDecoder() {
+        @Override
+        public Long decodeClientMessage(ClientMessage clientMessage) {
+            return RingbufferAddAllCodec.decodeResponse(clientMessage).response;
+        }
+    };
+
+    private ClientMessageDecoder readManyAsyncResponseDecoder;
+
+    private int partitionId;
+    private volatile long capacity = -1;
+
+    public ClientRingbufferProxy(String serviceName, String objectName) {
+        super(serviceName, objectName);
+    }
+
+    @Override
+    protected void onInitialize() {
+        partitionId = getContext().getPartitionService().getPartitionId(getName());
+        final SerializationService serializationService = getContext().getSerializationService();
+
+        readManyAsyncResponseDecoder = new ClientMessageDecoder() {
+            @Override
+            public PortableReadResultSet decodeClientMessage(ClientMessage clientMessage) {
+                Data response = RingbufferReadManyCodec.decodeResponse(clientMessage).response;
+                PortableReadResultSet readResultSet = serializationService.toObject(response);
```

Comment:
This usage does not seem to fit to client protocol. 
Only user specific data should come as Data in protocol. Anything primitive or structures encapsulating Data like List, Set , EntryView .. etc can be seen clearly in protocol.
You are gonna need to explain what is PortableResultSet in protocol. We used similar approaches in a couple of places. An example is EntryView 
https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ResponseTemplate.java#L76

https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ResponseTemplate.java#L76